### PR TITLE
Add RequestID field to Transactions

### PR DESCRIPTION
**Pull request by amit-momin**
[Proposal Doc](https://www.notion.so/chainlink/Untangle-TXM-CreateTransaction-from-other-components-143bb2b0a7f04581a86ad221974e6715)

- Added `request_id` column to the `eth_txes` table
- Updated logic for `CreateTransaction()`
  - If `RequestID` is null in TxRequest, proceed with `CreateTransaction()` as before
  - If `RequestID` is present  in TxRequest, check if Tx already existing in the `eth_txes` table with the same `RequestID`
    - If Tx exists, return the existing one
    - If Tx does not exist, proceed with `CreateTransaction()`

### CODE REVIEW

File: common/txmgr/txmgr.go [hunk 1]
```diff
@@ -462,6 +462,18 @@ func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CreateTran
...
+	// Skipping CreateTransaction to avoid double send
+	if txRequest.RequestID != nil {
+		etx, err := b.txStore.FindTxWithRequestID(*txRequest.RequestID)
+		if err != nil {
+			return tx, errors.Wrap(err, "Failed to search for transaction with RequestID")
+		}
+		if (etx != nil) {
+			return *etx, nil
+		}
+	}
```
Comment (samsondav) [inline common/txmgr/txmgr.go:475]: I think this check can be moved earlier in the function

File: core/chains/evm/txmgr/evm_tx_store.go [hunk 2]
```diff
@@ -906,6 +908,28 @@ func (o *evmTxStore) FindReceiptsPendingConfirmation(ctx context.Context, blockN
...
+	err = o.q.Transaction(func(tx pg.Queryer) error {
+		var dbEtx DbEthTx
+		err = tx.Get(&dbEtx, `SELECT * FROM eth_txes WHERE request_id = $1`, requestID)
+		if err != nil {
+			if !errors.Is(err, sql.ErrNoRows) {
+				return pkgerrors.Wrap(err, "FindTxWithRequestID failed to load eth_txes")
+			}
+			return nil
+		}	
+		DbEthTxToEthTx(dbEtx, etx)
```
Comment (samsondav) [inline core/chains/evm/txmgr/evm_tx_store.go:923]: This is new, why is this necessary? Can we not scan directly into the appropriate struct?

File: core/store/migrate/migrations/0191_add_request_id_column_eth_txes_table.sql [hunk 3]
```diff
@@ -0,0 +1,7 @@
+-- +goose Up
+
+ALTER TABLE eth_txes ADD COLUMN request_id bytea UNIQUE;
```
Comment (samsondav) [inline core/store/migrate/migrations/0191_add_request_id_column_eth_txes_table.sql:3]: Maybe worth adding a length limit as a sanity check e.g. 2000 bytes or something

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/chains/evm/txmgr/evm_tx_store.go (hunk 2 above)
Comment (amit-momin) [reply core/chains/evm/txmgr/evm_tx_store.go:923]: I ended up taking this from `FindTxWithSequence()` but looks like it has some specific logic around parsing the nonce column into the generic Tx type's sequence field.

File: common/txmgr/txmgr.go [hunk 4]
```diff
@@ -435,6 +435,19 @@ func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Trigger(ad
...
+	// Check for existing Tx with RequestID. If found, return the Tx and do nothing
+	// Skipping CreateTransaction to avoid double send
+	if txRequest.RequestID != nil {
+		var etx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+		etx, err = b.txStore.FindTxWithRequestID(*txRequest.RequestID)
+		if err != nil {
+			return tx, errors.Wrap(err, "Failed to search for transaction with RequestID")
+		}
+		if etx != nil {
+			return *etx, nil
```
Comment (prashantkumar1982) [inline common/txmgr/txmgr.go:447]: nit: Add an info log here, that this particular ReuqestId was found, so returning that, instead of creating a new tx.

File: common/txmgr/txmgr.go [hunk 5]
```diff
@@ -435,6 +435,19 @@ func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Trigger(ad
 
 // CreateTransaction inserts a new transaction
 func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CreateTransaction(txRequest txmgrtypes.TxRequest[ADDR, TX_HASH], qs ...pg.QOpt) (tx txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], err error) {
+	// Check for existing Tx with RequestID. If found, return the Tx and do nothing
+	// Skipping CreateTransaction to avoid double send
+	if txRequest.RequestID != nil {
+		var etx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
```
Comment (prashantkumar1982) [inline common/txmgr/txmgr.go:441]: nit: rename etx to existingTx or similar.
etx refers to old name, that implied Eth Tx, which no longer is correct, given this doesn't just represent ethereum/evm.

File: common/txmgr/types/tx.go [hunk 6]
```diff
@@ -66,6 +66,7 @@ func (s TxAttemptState) String() (str string) {
 }
 
 type TxRequest[ADDR types.Hashable, TX_HASH types.Hashable] struct {
+	RequestID        *TX_HASH
```
Comment (prashantkumar1982) [inline common/txmgr/types/tx.go:69]: Could you change this to string?
We just require clients to send us a unique string.
TX_HASH has a certain meaning, implying his field is kind of a TxHash. We don't actually care that this field be same as the TX_HASH type.

So this will be a string here, and also in the DB then.

Comment (prashantkumar1982) [inline common/txmgr/types/tx.go:69]: Since you are adding this field now, can you please add code comments indicating what this field means, nd how is this used by the TXM?

File: common/txmgr/types/tx_store.go [hunk 7]
```diff
@@ -64,6 +64,7 @@ type TransactionStore[
 	FindTxAttemptsConfirmedMissingReceipt(chainID CHAIN_ID) (attempts []TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], err error)
 	FindTxAttemptsRequiringReceiptFetch(chainID CHAIN_ID) (attempts []TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], err error)
 	FindTxAttemptsRequiringResend(olderThan time.Time, maxInFlightTransactions uint32, chainID CHAIN_ID, address ADDR) (attempts []TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], err error)
+	FindTxWithRequestID(requestID TX_HASH) (etx *Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], err error)
```
Comment (prashantkumar1982) [inline common/txmgr/types/tx_store.go:67]: nit: Add some code level comment here explaining what this function does.

File: core/chains/evm/txmgr/evm_tx_store.go [hunk 8]
```diff
@@ -152,6 +152,7 @@ func toOnchainReceipt(rs []*evmtypes.Receipt) []rawOnchainReceipt {
 // This is exported, as tests and other external code still directly reads DB using this schema.
 type DbEthTx struct {
 	ID             int64
+	RequestID      *common.Hash
```
Comment (prashantkumar1982) [inline core/chains/evm/txmgr/evm_tx_store.go:155]: so this too will be string if we change at all places

File: core/chains/evm/txmgr/evm_tx_store.go [hunk 9]
```diff
@@ -906,6 +908,25 @@ func (o *evmTxStore) FindReceiptsPendingConfirmation(ctx context.Context, blockN
...
+func (o *evmTxStore) FindTxWithRequestID(requestID common.Hash) (etx *Tx, err error) {
+	err = o.q.Transaction(func(tx pg.Queryer) error {
+		var dbEtx DbEthTx
+		err = tx.Get(&dbEtx, `SELECT * FROM eth_txes WHERE request_id = $1`, requestID)
+		if err != nil {
+			return pkgerrors.Wrap(err, "FindTxWithRequestID failed to load eth_txes")
+		}
+		etx = new(Tx)
+		DbEthTxToEthTx(dbEtx, etx)
+		err = o.LoadTxAttempts(etx, pg.WithQueryer(tx))
```
Comment (prashantkumar1982) [inline core/chains/evm/txmgr/evm_tx_store.go:921]: I don't think you need to make this call. We are just sending the existing Tx to caller, and it doesn't quite require loading the TxAttempts.

File: core/chains/evm/txmgr/evm_tx_store.go [hunk 10]
```diff
@@ -906,6 +908,25 @@ func (o *evmTxStore) FindReceiptsPendingConfirmation(ctx context.Context, blockN
 	return
 }
 
+// FindTxWithRequestID returns any broadcast ethtx with the given requestID
+func (o *evmTxStore) FindTxWithRequestID(requestID common.Hash) (etx *Tx, err error) {
+	err = o.q.Transaction(func(tx pg.Queryer) error {
+		var dbEtx DbEthTx
+		err = tx.Get(&dbEtx, `SELECT * FROM eth_txes WHERE request_id = $1`, requestID)
```
Comment (prashantkumar1982) [inline core/chains/evm/txmgr/evm_tx_store.go:915]: AND evm_chain_id = $2`, requestID, chain_id

Pass in chain_id from caller.

This is very important

File: core/chains/evm/txmgr/evm_tx_store.go [hunk 11]
```diff
@@ -906,6 +908,25 @@ func (o *evmTxStore) FindReceiptsPendingConfirmation(ctx context.Context, blockN
 	return
 }
 
+// FindTxWithRequestID returns any broadcast ethtx with the given requestID
+func (o *evmTxStore) FindTxWithRequestID(requestID common.Hash) (etx *Tx, err error) {
```
Comment (prashantkumar1982) [inline core/chains/evm/txmgr/evm_tx_store.go:912]: Instead of this function, you could directly add this logic inside CreateTransaction.
I see CreateTransaction already has this logic, for the field txRequest.PipelineTaskRunID.

It might be easier to just repeat that logic as it is, for your new requestId field.

**Review by dimriou — COMMENTED** [review]
I noticed that VRF already have a RequestID field [here](https://github.com/smartcontractkit/chainlink/blob/develop/common/txmgr/types/tx.go#L116). Is it referring to the same thing? Have we considered merging them?

File: common/txmgr/txmgr.go [hunk 12]
```diff
@@ -435,6 +435,19 @@ func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Trigger(ad
...
+		etx, err = b.txStore.FindTxWithRequestID(*txRequest.RequestID)
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (amit-momin) [reply common/txmgr/txmgr.go:443]: The `FindTxWithRequestID` method actually catches this error and returns nil if it does but I'll add this here too just in case

File: common/txmgr/txmgr.go [hunk 13]
```diff
@@ -435,6 +435,19 @@ func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Trigger(ad
 
 // CreateTransaction inserts a new transaction
 func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CreateTransaction(txRequest txmgrtypes.TxRequest[ADDR, TX_HASH], qs ...pg.QOpt) (tx txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], err error) {
+	// Check for existing Tx with RequestID. If found, return the Tx and do nothing
+	// Skipping CreateTransaction to avoid double send
+	if txRequest.RequestID != nil {
```
[the remark this answers is not in the document]
Comment (dimriou) [reply common/txmgr/txmgr.go:440]: Can you add a comment the RequestID is expected to be provided by VRF and Fluxmonitor, so we know that for other products is irrelevant?

[the remark this answers is not in the document]
Comment (amit-momin) [reply common/txmgr/txmgr.go:440]: That's the current expectation but don't know if other services may need it in the future. I can leave a comment that it's only used by "external services" if that gets the point across?

File: core/store/migrate/migrations/0191_add_request_id_column_eth_txes_table.sql [hunk 14]
```diff
@@ -0,0 +1,7 @@
+-- +goose Up
+
+ALTER TABLE eth_txes ADD COLUMN request_id bytea UNIQUE CHECK (octet_length(request_id) <= 2000);
```
[the remark this answers is not in the document]
Comment (amit-momin) [reply core/store/migrate/migrations/0191_add_request_id_column_eth_txes_table.sql:3]: Yes that's right. Postgres doesn't treat nulls as equal unless specified so we can have the UNIQUE constraint with multiple null values. Looks like we don't need to separately index the new column either since the UNIQUE constraint automatically takes care of that. https://www.postgresql.org/docs/current/indexes-unique.html

File: core/chains/evm/txmgr/evm_tx_store.go (hunk 11 above)
Comment (amit-momin) [reply core/chains/evm/txmgr/evm_tx_store.go:912]: I was considering that originally but thought it'd be better to maintain this logic in the TXM `CreateTransaction` so it's generalized. Otherwise when creating a new TxStore, the developer would be expected to include this logic. If that's not something we don't need to worry about, I can move this there.

**Review by prashantkumar1982 — APPROVED** [review]
Thanks, looks good now!

File: common/txmgr/txmgr.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (dimriou) [reply common/txmgr/txmgr.go:441]: Yeah, that works.

File: core/store/migrate/migrations/0191_add_request_id_column_eth_txes_table.sql [hunk 15]
```diff
@@ -0,0 +1,7 @@
+-- +goose Up
+
+ALTER TABLE eth_txes ADD COLUMN request_id varchar(2000) UNIQUE;
```
Comment (samsondav) [inline core/store/migrate/migrations/0191_add_request_id_column_eth_txes_table.sql:3]: Assuming that `UNIQUE` constraint automatically generates an index?

**Review by samsondav — APPROVED** [review]

**Review by prashantkumar1982 — APPROVED** [review]

**Review by samsondav — APPROVED** [review]

File: common/txmgr/types/tx.go [hunk 16]
```diff
@@ -66,6 +66,10 @@ func (s TxAttemptState) String() (str string) {
 }
 
 type TxRequest[ADDR types.Hashable, TX_HASH types.Hashable] struct {
+	// Unique ID set by external components to prevent double send
+	// If present in TxRequest, TXM will search for Tx with ID
+	// If found, it will return the existing Tx, otherwise it will create a new Tx
+	RequestID        *string
```
Comment (jmank88) [inline common/txmgr/types/tx.go:72]: Can we try to think of a name for this field that doesn't result in stuttering? `TxRequest.RequestID`

Comment (jmank88) [reply common/txmgr/types/tx.go:72]: > ...TXM will search for Tx with ID
> // If found, it will return the existing Tx

What does 'return' mean in this context? Can we describe the effect here without including any implementation details? What if rest of the fields are different? Is that info discarded, or is the existing one updated?

Comment (prashantkumar1982) [reply common/txmgr/types/tx.go:72]: Don't have strong opinion on the name. RequestId was already used in the TxMeta field by VRF, so made sense to reuse the same name. Is stuttering a real problem? I don't see people having to read out the whole thing.
I could think of other alternatives:
- UniqueId
- IdempotencyKey

For the comment on this field, how about changing to:
// RequestId is a globally unique ID set by the caller, to prevent accidental creation of duplicated Txs during retries or crash recovery.
// If this field is set, the TXM will first search existing Txs with this field.
// If found, it will return the existing Tx, without creating a new one. TXM will not validate or ensure that existing Tx is same as the incoming TxRequest.
// If not found, TXM will create a new Tx.
// If requestID is set to null, TXM will always create a new Tx.
// Callers must ensure that they use a globally unique id while setting this field.

Comment (jmank88) [reply common/txmgr/types/tx.go:72]: > Is stuttering a real problem?

Eliminating redundant words is one of the few pieces of advice on naming: https://go.dev/doc/effective_go#names

In this case, I think it reveals that there is more to say due to the overloaded terminology obscuring things. A different word like `Idempotency` definitely adds value here :+1: 

> // Callers must ensure that they use a globally unique id while setting this field.

Do we need to protect them from messing this up? e.g. by using a `uuid.UUID`

Comment (amit-momin) [reply common/txmgr/types/tx.go:72]: I'll go ahead and change it to `IdempotencyKey` then and add the new comment above. I wouldn't be opposed to changing it to `uuid` either. It'd ensure unique values are being used across services that need this.

**Review by jmank88 — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]

**Review by prashantkumar1982 — APPROVED** [review]
